zerver/lib/attachments.py

```diff
@@ -160,6 +162,24 @@ def validate_attachment_request(
         # can access it.
         return True, attachment
 
+    message_channel_ids = set()
+    for message in messages:
+        if message.is_stream_message():
+            message_channel_ids.add(message.recipient.type_id)
+
+    # We only fetch channels with public history, since those are the
+    # only channels that user groups can grant content access to.
+    message_channels_with_public_history = Stream.objects.filter(
+        id__in=message_channel_ids, history_public_to_subscribers=True
+    )
+    if len(message_channels_with_public_history) > 0:
+        user_recursive_group_ids = set(
+            get_recursive_membership_groups(user_profile).values_list("id", flat=True)
+        )
+        for channel in message_channels_with_public_history:
+            if is_user_in_groups_granting_content_access(channel, user_recursive_group_ids):
+                return True, attachment
+
```

Comment:
Accessing uploaded files is a very hot code path, so I think it is probably worth more attention to optimize it. Here, you're treating "subscriber" differently from "has content" access in a way that probably results in duplicated work.

I think probably instead we want to just make the `relevant_channel_ids` calculation include channels accessible via content access, not just active Subscriptions.